### ISSUE TITLE
[prefer-readonly-parameter-types] False positive with function-typed argument

### ISSUE THREAD

**Issue by bbrk24** [typescript-eslint/typescript-eslint#4752]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": [
      "error",
      {
        "ignoreInferredTypes": true
      }
    ]
  }
}
```

```TS
function testfunc(callback: (arg: Object) => void) {
}
```

```JSON
{
    "include": ["*.ts"],
    "compilerOptions": {
        "allowUnusedLabels": false,
        "exactOptionalPropertyTypes": true,
        "noImplicitReturns": true,
        "noUncheckedIndexedAccess": true,
        "strict": true,
        "target": "ES2017",
        "lib": ["ES2017", "DOM"]
    }
}
```

**Expected Result**

`prefer-readonly-parameter-types` should not trigger, since there are no mutable arguments to the function.

**Actual Result**

The rule triggers on the `Object`-typed parameter of the callback, since it's a mutable argument to a function type.

**Additional Info**

Part of the output from running `eslint`with `--debug`:

```
  eslintrc:ignore-pattern Check {
  filePath: 'C:\\[...]\\src\\index.ts',
  dot: false,
  relativePath: 'src/index.ts',
  result: false
} +1s
  eslint:file-enumerator Yield: index.ts +1s
  eslintrc:cascading-config-array-factory Load config files for C:\[...]\src. +1s
  eslintrc:cascading-config-array-factory Cache hit: C:\[...]\src. +0ms
  eslint:cli-engine Lint C:\[...]\src\index.ts +1s
  eslint:linter Linting code for C:\[...]\src\index.ts (pass 1) +5ms
  eslint:linter Verify +0ms
  eslint:linter With ConfigArray: C:\[...]\src\index.ts +0ms
  eslint:linter Parsing: C:\[...]\src\index.ts +1ms
  eslint:linter Parsing successful: C:\[...]\src\index.ts +41ms
  eslint:linter Scope analysis: C:\[...]\src\index.ts +1ms
  eslint:linter Scope analysis successful: C:\[...]\src\index.ts +0ms
  eslint:linter Generating fixed text for C:\[...]\src\index.ts (pass 1) +191ms
  eslint:source-code-fixer Applying fixes +238ms
  eslint:source-code-fixer shouldFix parameter was false, not attempting fixes +0ms
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.16.0` |
| `@typescript-eslint/parser`        | `5.16.0` |
| `TypeScript`                       | `4.6.3` |
| `ESLint`                           | `8.12.0` |
| `node`                             | `14.17.0` |

This started happening when I upgraded typescript-eslint from v5.9 to v5.10 (at the time, I was using typescript 4.4.3 and eslint 8.9.0). The issue occurs in every version of typescript-eslint since that I have tested.

**Comment by bradzacher** [comment]
This is working as expected - the rule checks *all* functions - real or types.
It always has - [you can see its handling in the very first version of the rule that was released back in v2.22](https://github.com/typescript-eslint/typescript-eslint/commit/3be98542afd7553cbbec66c4be215173d7f7ffcf#diff-3ebbf6cc405d2556a7c1860f7f2ebba6a1e6e3dbd7c0f3bb19dd29434036444eR58).

> at the time, I was using typescript 4.4.3

Likely this was exposed by upgrading your TS version.

**Comment by bbrk24** [comment]
@bradzacher That is not the case. Having tested this with various combinations of package versions, this isn't happening with TS 4.5 and typescript-eslint 5.9, and it is happening with TS 4.5 and typescript-eslint >=5.10.
